flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java

```diff
@@ -0,0 +1,812 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.isProctimeIndicatorType;
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.isRowtimeIndicatorType;
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.isTimeIndicatorType;
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.isTimestampLtzIndicatorType;
+import static org.apache.flink.table.planner.plan.utils.MatchUtil.isFinalOnMatchTimeIndicator;
+import static org.apache.flink.table.planner.plan.utils.MatchUtil.isMatchRowTimeIndicator;
+import static org.apache.flink.table.planner.plan.utils.WindowUtil.groupingContainsWindowStartEnd;
+
+/**
+ * Traverses a {@link RelNode} tree and converts fields with {@link TimeIndicatorRelDataType} type.
+ * If a time attribute is accessed for a calculation, it will be materialized. Forwarding is allowed
+ * in some cases, but not all.
+ */
+public final class RelTimeIndicatorConverter implements RelShuttle {
+
+    private final RexBuilder rexBuilder;
+
+    private RelTimeIndicatorConverter(RexBuilder rexBuilder) {
+        this.rexBuilder = rexBuilder;
+    }
+
+    public static RelNode convert(
+            RelNode rootRel, RexBuilder rexBuilder, boolean needFinalTimeIndicatorConversion) {
+        RelTimeIndicatorConverter converter = new RelTimeIndicatorConverter(rexBuilder);
+        RelNode convertedRoot = rootRel.accept(converter);
+
+        // LogicalLegacySink and LogicalSink are already converted
+        if (rootRel instanceof LogicalLegacySink
+                || rootRel instanceof LogicalSink
+                || !needFinalTimeIndicatorConversion) {
+            return convertedRoot;
+        }
+        // materialize remaining procTime indicators
+        return converter.materializeProcTime(convertedRoot);
+    }
+
+    @Override
+    public RelNode visit(LogicalIntersect intersect) {
+        return visitSetOp(intersect);
+    }
+
+    @Override
+    public RelNode visit(LogicalUnion union) {
+        return visitSetOp(union);
+    }
+
+    @Override
+    public RelNode visit(LogicalMinus minus) {
+        return visitSetOp(minus);
+    }
+
+    @Override
+    public RelNode visit(LogicalAggregate aggregate) {
+        return visitAggregate(aggregate);
+    }
+
+    @Override
+    public RelNode visit(LogicalSort sort) {
+        return visitSimpleRel(sort);
+    }
+
+    @Override
+    public RelNode visit(LogicalMatch match) {
+        RelNode newInput = match.getInput().accept(this);
+        RexTimeIndicatorMaterializer materializer = new RexTimeIndicatorMaterializer(newInput);
```

Comment:
the logic is changed: the previous behavior is using the unconverted input type, while the current behavior is using the converted input type.

Reply:
Yes, the previous behavior is wrong. but it is a little hard when I try to find a case to cover this, so I skip the test case. I would try harder to find a case....

Reply:
@godfreyhe, add a test `MatchRecognizeTest#testCascadeMatch` to cover this.